.crawler/src/metrics.rs

```diff
@@ -0,0 +1,416 @@
+        }
+
+        writeln!(f, "\nBreakdown:")?;
+        write!(f, "by node type: ")?;
+        print_breakdown(f, &self.types, 0)?;
+        write!(f, "by protocol version: ")?;
+        print_breakdown(f, &self.versions, 0)?;
+        write!(f, "by current state: ")?;
+        print_breakdown(f, &self.states, 0)?;
+
+        writeln!(f, "\nNetwork metrics:")?;
+        writeln!(f, "density: {:.4}", self.density)?;
+        writeln!(f, "algebraic connectivity: {:.4}", self.algebraic_connectivity)?;
+        writeln!(f, "degree centrality delta: {}", self.degree_centrality_delta)?;
+        writeln!(f, "average number of node connections: {}", self.avg_degree_centrality)?;
+        if let Some(ms) = self.avg_handshake_time_ms {
+            writeln!(f, "average handshake time: {}ms", ms)?;
+        }
+
+        writeln!(f, "\nBlockchain-related details:")?;
+        if let Some(h) = self.heights.keys().max() {
+            writeln!(f, "maximum found height: {}", h)?;
+        }
+        write!(f, "5 most common heights: ")?;
+        print_breakdown(f, &self.heights, 5)?;
+        if let Some(h) = self.avg_height {
+            writeln!(f, "average height: {}", h)?;
+        }
+
+        Ok(())
+    }
+}
+
+/// Network topology measurements.
+#[derive(Debug)]
+pub struct NetworkMetrics {
+    /// The total node count of the network.
+    pub node_count: usize,
+    /// The total connection count for the network.
+    pub connection_count: usize,
+    /// The network density.
+    ///
+    /// This is defined as actual connections divided by the total number of possible connections.
+    pub density: f64,
+    /// The algebraic connectivity of the network.
+    ///
+    /// This is the value of the Fiedler eigenvalue, the second-smallest eigenvalue of the network's
+    /// Laplacian matrix.
+    pub algebraic_connectivity: f64,
+    /// The difference between the node with the largest connection count and the node with the
+    /// lowest.
+    pub degree_centrality_delta: u16,
+    /// Per-node metrics.
+    pub per_node: Vec<(SocketAddr, NodeMeta, NodeCentrality)>,
+}
+
+impl NetworkMetrics {
+    /// Returns the network metrics for the state described by the connections list.
+    pub fn new(connections: HashSet<Connection>, nodes: HashMap<SocketAddr, NodeMeta>) -> Option<Self> {
+        // Don't compute the metrics for an empty set of connections.
+        if connections.is_empty() {
+            // Returns all metrics set to `0`.
+            return None;
+        }
+
+        let node_count = nodes.len();
+        let connection_count = connections.len();
+        let density = calculate_density(node_count as f64, connection_count as f64);
+
+        // Create an index of nodes to introduce some notion of order the rows and columns all matrices will follow.
+        let index: BTreeMap<SocketAddr, usize> = nodes.keys().enumerate().map(|(i, addr)| (*addr, i)).collect();
```

Comment:
We'll want this index to be sorted—order preservation is very helpful for debugging and correlating data over multiple sets of these measurements.

Reply:
You mean sorted as in a more specified order? As this list already has a predictable order (i.e. is somewhat sorted), which depends on `addr`.

Reply:
Actually, the ordering of `SocketAddr` seems sufficient here 👍